### ISSUE TITLE
[Elixir] Add support for Elixir and PCRE

### ISSUE THREAD

**Issue by azizk** [sublimehq/Packages#2825]
Hi there!

I'm proud to present a [high-quality syntax package](https://github.com/princemaple/elixir-sublime-syntax) for the Elixir programming language. The focus was on completeness and accuracy of the syntax definition as well as its visual appeal. It's got an extensive test-suite and color scheme adaptations for Mariana and Monokai and countless other niceties.

There's even a complete definition for the syntax of [PCRE](https://www.pcre.org), although with a focus on the [regexp documentation](https://erlang.org/doc/man/re.html) provided by Erlang. So this file could practically be used in Erlang's definition as well (for example, by matching on particular functions from the `re` module).

Currently at least two packages exist for Elixir. The creator of Elixir, @josevalim, maintains his own package that was adapted from TextMate's Ruby package. I received his official approval for my endeavour to get my work accepted as *the* default Elixir syntax package in Sublime Text: https://github.com/elixir-editors/elixir-tmbundle/issues/188#issuecomment-843816749

Existing packages won't necessarily lose their relevance, because they still provide other things such as completions, snippets etc. But they should probably be modified in order to not confuse the editor about which definition is to be used.

I know you mention in the readme that you don't intend to add new packages, but I'm hopeful that you will consider my application nonetheless.

**Comment by deathaxe** [comment]
I can't say something about acceptence of new syntaxes. Providing it as 3rd-party package via Package Control gives you more flexibility with regards to update cycles. You can ship fixes faster, which might be an advantage.

Your syntax makes use of features available to Sublime Text 4 only.

If the original package _elixir-editors/elixir-tmbundle_ wants to keep TextMate compatibility it can't make use of full power of ST's syntax engine. So if your package provides at least the same set of features it might replace the original package for ST4 and higher - eighter by re-registering the package name to your repo or maintaining your changes in a separate branch of the original repository.

The final result was the same. Only one Elexir package on Package Control.

Several packages go this way to create st4 branches (or freeze current state in st3 branches to keep backward compatibility).

I am about to prepare it for [MarkdownEditing](https://github.com/SublimeText-Markdown/MarkdownEditing/issues/602).
Another example is https://github.com/SublimeText/PowerShell

The only thing you'd need to do is setup proper tag names and version boundaries in package_control_channel.

**Comment by azizk** [comment]
Thanks for explaining!

You're right, it's fast to publish changes if it remains a package, but I'm not too worried about that. It's well-tested and it's fine if some fixes take a bit time to get deployed. One of the advantages I hope to see is to be able to use/see Elixir inside Markdown code blocks. As a neat side-effect I'd hope to get syntax highlighted Elixir code inside info pop-ups provided by the language server (I assume they're rendered using Markdown.) Right now it's just boring white text. OK, after writing this I solved the first issue for myself by overriding Markdown.sublime-syntax, but still...

I'm only using the syntax extension feature of ST4. We actually have a branch which keeps ST3 compatibility with a few differences here and there. Afaics @princemaple (package maintainer) already got the tag and version settings into PCC for both ST3 and ST4 a while ago. :+1:

One package to bundle them all sounds like a great idea, but José expressed his wish of keeping TextMate compatibility and that's alright. So I'm not sure. Would still be cool to have Elixir and PCRE directly in Sublime Text. :grinning:

**Comment by keith-hall** [comment]
PCRE is already supported - https://github.com/sublimehq/Packages/blob/3eb904992f6409c186c9524afea66d56002f73af/Regular%20Expressions/RegExp.sublime-syntax#L4

though I've noticed the `\Q...\E{1,2}` support needs some fixes, I'll submit a PR for that soon

**Comment by keith-hall** [comment]
> though I've noticed the `\Q...\E{1,2}` support needs some fixes, I'll submit a PR for that soon

fixed in https://github.com/sublimehq/Packages/pull/2836

**Comment by azizk** [comment]
You're right, there is already support for it. I think when I looked at it 2 years ago I found it wasn't really complete or lacking in some ways, so I decided to implement my own syntax definition for PCRE from scratch. I'm not sure how they compare now, but I'll take a look. Good to see you fixed the quoted literal feature! :+1:

**Comment by deathaxe** [comment]
I just had another look at the proposed syntax package as a user wanted to use it for "Surface" syntax.

The proposed Elixir syntax package makes heavy use of 

1. "lookbehinds", which cause ST to fallback to Oniguruma regex engine, which is much slower then the primarily used SRegex engine.
2. "with_prototype", which is known to cause inclusion-loops in various situations. The result is the widely known "25.000" context sanity limit error.

In short: The proposed syntax is not in a shape to be added to core syntax.

**Comment by azizk** [comment]
Thank you for taking the time and the explanation!

Does this mean, that if I made a version available without the mentioned issues, you would commit and ship the syntax definition?
I didn't know these criteria existed. Are there any others? Is there some documentation somewhere? A guide would definitely be helpful!

Regarding point 1., I really don't see how that's an issue. Of course, I don't know how much faster the highlighter would be without look-behinds, but I've edited and looked at copious amounts of Elixir code with my syntax definition for countless hours.
It's blazingly fast! Speed is important to me and that's the main reason I use Sublime Text. Even if I generate an enormous file with over 100_000 lines, ST takes around 8s on my i7-4710HQ with 4 cores to open and render the file. But nobody will ever edit a file like that... Besides, a few integrated syntax files use look-behinds like AppleScript and OCaml.

Point 2. is, on the other hand, a tough problem and often a huge pain to work around. If I were to create a sanitized version, instead of removing all the look-behinds, I'd rather remove as many `with_prototype` instances as possible. I'm counting around 33 occurrences. This stems mostly from the variations of string delimiters and syntax interpolations. (I'm curious about the context limit error: is there a definitive discussion about it and whether it will ever be fixed?)

Okay, it's not specifically because of your reply, but when I think about it now it's fine if my work isn't shipped with ST. That was my desire initially but at the moment I don't feel like it's a goal I want to achieve so much anymore.
Maybe I'll try again later. Maybe someone else will...

We can still discuss, but I'll be a good citizen and close this issue for now. Thanks for your attention!

**Comment by deathaxe** [comment]
I can't influence what's added and what not. Besides "Swift" and maybe "Kotlin" I am not aware of any plans to add further syntaxes. I just wanted to point out the two major flaws I found in this syntax while guiding a user how to extend it to support "Surface" templates. The `with_prototype` issue makes it literly impossible.

> Regarding point 1., I really don't see how that's an issue. 

Oniguruma is triggered for backward compatibility only. The goal is to minimize situations it is used. I don't think it will be dropped anytime soon, but sregex was created on purpose.

Much effort was put into default syntaxes to remove all the lookbehinds. All syntaxes I rewrote to drop those got 20-40% faster. It may depend on several other syntax architectural factors, too, though.

Please don't misunderstand my critics! The overall syntax definition looks quite well and it is clearly noticible much effort and love was put into.

>  a few integrated syntax files use look-behinds like AppleScript and OCaml.

Yes, because those are terribly outdated anyway. Noone cared about them much since they have been converted from tmLanguage to sublime-syntax in 2015.

They can't be used for benchmarking at all. Neighter coding style, scope names, accuracy nor anything. Some rough edges have been addressed, but much are still pending.

> it's fine if my work isn't shipped with ST

It's very much ok to have well maintained and sophisticated packages/plugins available. As stated earlier it's more flexible for you to do updates or fix bugs. It helps to keep ST lean for those who don't need that syntax.

> (I'm curious about the context limit error: is there a definitive discussion about it and whether it will ever be fixed?)

According to some discussions during ST4 dev cycle it was said to be addressed in various ways or even fixed.

I guess the issue is including a syntax into itself via third party syntax, which is hard or impossible to detect and prevent recursions for.

If I create another string context in your Elixir syntax, which includes `main` context, everything seems fine. As soon as I include the Surface syntax (html with embedded Elixir), sanity limit is hit. 

I can't judge, whether something like that is fixable. That's something core devs would have to look into.

**Comment by azizk** [comment]
That's great info on the Oniguruma look-behind issue. I didn't know it could affect the speed that much. However, for now I won't make efforts to eliminate them, because it would be hard and time-intensive to make everything work and I'd probably have to use the branching feature several times.

> Please don't misunderstand my critics! The overall syntax definition looks quite well and it is clearly noticible much effort and love was put into.

No worries! I didn't take this as criticism really. I'm not disappointed or anything like that. You're right that it's totally fine to have it exist as an external package which is better for closer control and maintenance.

> According to some discussions during ST4 dev cycle it was said to be addressed in various ways or even fixed.

I've encountered the error many times while I was working on my Elixir syntax definition for ST4. I suppose they improved certain cases but it's far from being fixed completely... I wish they would put more time into fixing it for good.

So I just published [a new release](/princemaple/elixir-sublime-syntax/pull/27) of the Elixir package and I'm actually planning on supporting the Surface syntax. Is it really impossible? I was thinking about copying "HTML (EEx).sublime-syntax" as a template. It can highlight Elixir code tags inside HTML. I just tested this in fact (e.g. `<div class={@my_class}>`) and it looks like this is going to work. Hopefully there won't be any troubles further ahead, but it doesn't seem impossible atm. :+1:

> If I create another string context in your Elixir syntax, which includes main context, everything seems fine. As soon as I include the Surface syntax (html with embedded Elixir), sanity limit is hit.

The trick is to comment out one of the other string contexts which embeds another syntax. I'll have to make a compromise but I hope I can simply replace the `~L""` contexts which will become obsolete anyway.

**Comment by deathaxe** [comment]
Surface syntax itself seems easy. Created a gist for it https://gist.github.com/deathaxe/dabcd7575951174d08882b46842dd44a.

What we were failing with is embedding/including it back into Elixir in `~F"<surface>"` strings.

Maybe the trick is to exclude something from the Elixir used for Surface or so. Not sure. Don't have enough insight in Elixir's architecture to find a solution.

It's great news to here you planning to support it. It's the best solution, IMHO.

**Comment by azizk** [comment]
Thanks, I think it's great you tried to help that user. I wonder why they didn't open an issue ticket on our tracker. :thinking:

Thanks for sharing your syntax file. I tried it out. Afaics only `<#slot>`,  `<#template>` and  `<#Raw>` are missing. Not sure, because I have to give the Surface docs a thorough read as I haven't used it before.

Here is my [Surface ticket](/princemaple/elixir-sublime-syntax/issues/28) I just created. :+1:

**Comment by azizk** [comment]
Hey @deathaxe, hope you're doing fine! I finished the new syntax files for HEEx and Surface templates. I'd appreciate it, if you could spare some time to review these files. Would love to know if there's anything that could be done better. :)
